### ISSUE TITLE
[help wanted]LightGBM: Sklearn and Native API equivalence

### ISSUE THREAD

**Issue by dokteurwho** [Microsoft/LightGBM#1033]
I'm experimenting LightGBM through Training API <http://lightgbm.readthedocs.io/en/latest/Python-API.html#training-api> and Scikit-learn API <http://lightgbm.readthedocs.io/en/latest/Python-API.html#scikit-learn-api>.

I've not been able to make a clear mapping between both API as highlighted in example below. The basic idea is to train on 50% of the synthetic dataset.

    import numpy as np
    import lightgbm as lgbm
    
    # Generate Data Set
    xs = np.linspace(0, 10, 100).reshape((-1, 1)) 
    ys = xs**2 + 4*xs + 5.2
    ys = ys.reshape((-1,))
    
    # LGBM configuration
    alg_conf = {
        "num_boost_round":25,
        "max_depth" : 3,
        "num_leaves" : 31,
        'learning_rate' : 0.1,
        'boosting_type' : 'gbdt',
        'objective' : 'regression_l2',
        "early_stopping_rounds": None,
    }
    
    # Calling Regressor using scikit-learn API 
    sk_reg = lgbm.sklearn.LGBMRegressor(
        num_leaves=alg_conf["num_leaves"], 
        n_estimators=alg_conf["num_boost_round"], 
        max_depth=alg_conf["max_depth"],
        learning_rate=alg_conf["learning_rate"],
        objective=alg_conf["objective"]
    )
    sk_reg.fit(xs[::2], ys[::2])
    
    print("Scikit-learn API results")
    print(sk_reg.predict(xs[1::2]))
    
    
    # Calling Regressor using native API 
    train_dataset = lgbm.Dataset(xs[::2], ys[::2])
    lg_reg = lgbm.train(alg_conf.copy(), train_dataset)
    
    print("Native API results")
    print(lg_reg.predict(xs[1::2]))

# Output

    Scikit-learn API results
    [  14.35693851   14.35693851   14.35693851   14.35693851   14.35693851
       14.35693851   14.35693851   14.35693851   14.35693851   14.35693851
       25.37944751   25.37944751   25.37944751   25.37944751   25.37944751
       35.10572544   35.10572544   35.10572544   35.10572544   35.10572544
       46.50667974   46.50667974   46.50667974   46.50667974   46.50667974
       59.44952419   59.44952419   59.44952419   59.44952419   59.44952419
       75.42846332   75.42846332   75.42846332   75.42846332   75.42846332
      109.4610814   109.4610814   109.4610814   109.4610814   109.4610814
      109.4610814   109.4610814   109.4610814   109.4610814   109.4610814
      109.4610814   109.4610814   109.4610814   109.4610814   109.4610814 ]
    Native API results
    [ 22.55947971  22.55947971  22.55947971  22.55947971  22.55947971
      22.55947971  22.55947971  22.55947971  22.55947971  22.55947971
      22.55947971  22.55947971  22.55947971  22.55947971  22.55947971
      22.55947971  22.55947971  22.55947971  22.55947971  22.55947971
      45.33537795  45.33537795  45.33537795  45.33537795  45.33537795
      91.6376959   91.6376959   91.6376959   91.6376959   91.6376959
      91.6376959   91.6376959   91.6376959   91.6376959   91.6376959
      91.6376959   91.6376959   91.6376959   91.6376959   91.6376959
      91.6376959   91.6376959   91.6376959   91.6376959   91.6376959
      91.6376959   91.6376959   91.6376959   91.6376959   91.6376959 ]

# Question

Where could I find a clear equivalence between both API parameters ?

Thanks a lot.

**Comment by StrikerRUS** [comment]
@dokteurwho All parameters you could use in both cases are listed here: http://lightgbm.readthedocs.io/en/latest/Parameters.html

Try to set `n_estimators` instead of `num_boost_round` in Scikit-learn API.
This is one of the specific param name in sklearn (like `random_state`, `n_jobs`, `silent`).

Also take a look to log and make sure there are no records like
`Warning: "Param is set with Param=x, Param=y will be ignored. Current value: Param=x."` or 
`Warning: "Param is set with Param=y, will be overrided by Param=x. Current value: Param=x."`

And may I ask what's the real use case which requires to mix up using two APIs?

**Comment by dokteurwho** [comment]
Thanks for your feedback.

I'm using the `n_estimators ` parameter is Scikit-learn API. It looks to be consistent. See:
`
sk_reg = lgbm.sklearn.LGBMRegressor(
    ...,
    n_estimators=alg_conf["num_boost_round"], 
    ...
`

I didn't observe any log similar to the warning you quoted.

I want to move my model to Scikit-learn API.  In this perspective I want to process a non regression on my current model.

**Comment by StrikerRUS** [comment]
Oh, I see.

Please try to use the latest code from master branch. There was an occurrence of inconsistent parameters in   Scikit-learn API which was fixed yesterday: #1018.

**Comment by wxchan** [comment]
I think current master works fine now as long as you don't use random seed.

**Comment by dokteurwho** [comment]
Thank you ! With this code (ie adding alg_conf "min_child_weight": 1e-3, "min_child_samples": 20):

```
import numpy as np
import lightgbm as lgbm

# Generate Data Set
xs = np.linspace(0, 10, 100).reshape((-1, 1)) 
ys = xs**2 + 4*xs + 5.2
ys = ys.reshape((-1,))

# Or you could add to your alg_conf "min_child_weight": 1e-3, "min_child_samples": 20.

# LGBM configuration
alg_conf = {
    "num_boost_round":25,
    "max_depth" : 3,
    "num_leaves" : 31,
    'learning_rate' : 0.1,
    'boosting_type' : 'gbdt',
    'objective' : 'regression_l2',
    "early_stopping_rounds": None,
    "min_child_weight": 1e-3, 
    "min_child_samples": 20
}

# Calling Regressor using scikit-learn API 
sk_reg = lgbm.sklearn.LGBMRegressor(
    num_leaves=alg_conf["num_leaves"], 
    n_estimators=alg_conf["num_boost_round"], 
    max_depth=alg_conf["max_depth"],
    learning_rate=alg_conf["learning_rate"],
    objective=alg_conf["objective"],
    min_sum_hessian_in_leaf=alg_conf["min_child_weight"],
    min_data_in_leaf=alg_conf["min_child_samples"]
)
sk_reg.fit(xs[::2], ys[::2])

print("Scikit-learn API results")
print(sk_reg.predict(xs[1::2]))


# Calling Regressor using native API 
train_dataset = lgbm.Dataset(xs[::2], ys[::2])
lg_reg = lgbm.train(alg_conf.copy(), train_dataset)

print("Native API results")
print(lg_reg.predict(xs[1::2]))

```
It works fine.
